docs_src/design/adr/security/0019-spiffe.md

```diff
@@ -0,0 +1,158 @@
+
+### Technical Architecture
+
+![SPIFFE Architecture and Workflow](0019-spiffe-architecture.jpg)
+
+The work flow is as follows:
+
+1. The SPIFFE server is started.
+1. The server creates an initial set of keys.
+1. The trust bundle (certificate authority) data is exported from the SPIFFE server
+   and workload entries are loaded into the SPIFFE server database.
+1. A join token for the SPIFFE agent is created.
+1. The SPIFFE agent is started with the join token to add it to the cluster.
+1. Vault is started and `security-secret-store-setup`
+   initializes it and creates an admin token for `security-spiffe-token-provider` to use.
+1. The `security-spiffe-token-provider` service is started.
+   It obtains an SVID from the SIFFE agent and uses it as a TLS server certificate.
+1. An EdgeX microservice starts and obtains another SVID from the SPIFFE agent
+   and uses it as a TLS client certificate to contact the
+   `security-spiffe-token-provider` service.
+   The EdgeX microservice uses the trust bundle as a server CA
+   to verify the TLS certificate of the remote service.
+1. `security-spiffe-token-provider` verifies the SVID using the trust bundle as client CA
+   to verify the client,
+   extracts the service key,
+   and issues an appropriate Vault service token.
+1. The EdgeX microservice accesses Vault as usual.
+
+## Consequences
+
+This proposal will require addition of several new, optional, EdgeX microservices:
+
+* `security-spiffe-token-provider`, running on the main node
+* `spiffe-agent`, running on the main node and each remote node
+* `spiffe-server`, running on the main node
+* `spiffe-config`, a one-shot service running on the main node
+
+Minor changes will be needed to `security-secretstore-setup` to preserve
+the token-creating-token used by `security-file-token-provider` so that
+it can be used by `security-spiffe-token-provider`.
+
+The startup flow of the framework will be adjusted as follows:
+
+* Bootstrap service (original)
+  * `spiffe-server`
+    * `spiffe-config` (can be combined with `spifee-server`)
+      * `spiffe-agent`
+  * Vault service (original)
+    * Secret store setup service (original)
+      * `security-spiffe-token-provider`
+    * Consul (original)
+    * Postgres (original)
+
+There is no direct dependency between `spiffe-server`
+and any other microservice.
+`security-spiffe-token-provider` requires an SVID
+from `spiffe-agent` and a Vault admin token.
+
+None of these new services will be proxied via the API gateway.
+
+A new configuration file will be added to `spiffe-config`
+to configure the workload authorization parameters.
+This configuration is dependent upon whether one is running
+in a Snap configuration or a Docker configuration.
+For Snap configurations, new services with be authenticated
+using the Unix file path of the executable.
+For Docker configurations, the new services will be authenticated
+using the Docker service name (coded into the running container as a label).
```

Comment:
An example of this would be:
```
docker:label:com.docker.compose.service:edgex-core-data
```

In the case of Docker, you have to be root or member of the docker group to control the Docker daemon.  This threat actor is out of scope for EdgeX as we can't defend against an attacker with control of the container runtime.

In the case of a Snap, paths are not namespaced, so the `unix:path:/path/to/service/exe` selector, will point to a service executable in an installed snap.  If that is not sufficient protection, there is also a `unix:sha256` option that will hash the service executable.

In both cases, Unix file permissions control access to the SIFFE server API and database.  One need only bypass file system ACLS, connect to the SPIFFE server, and authorize additional workloads.  This is not any weaker than the existing Snap protections.

Reply:
When/where is this label added to running container?

Reply:
It is done automatically by docker-compose when launching a container.

Reply:
Nice!